controller/api/destination/server.go

```diff
@@ -383,11 +427,38 @@ func getSvcID(k8sAPI *k8s.API, clusterIP string, log *logging.Entry) (*watcher.S
 	return service, nil
 }
 
-// getPod returns a pod that maps to the given IP address. The pod can either
+// getPodByHostname returns a pod that maps to the given hostname (or an
+// instanceID). The hostname is generally the prefix of the pod's DNS name;
+// since it may be arbitrary we need to look at the corresponding service's
+// Endpoints object to see whether the hostname matches a pod.
+func getPodByHostname(k8sAPI *k8s.API, hostname string, svcID watcher.ServiceID) (*corev1.Pod, error) {
+	ep, err := k8sAPI.Endpoint().Lister().Endpoints(svcID.Namespace).Get(svcID.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, subset := range ep.Subsets {
+		for _, addr := range subset.Addresses {
+			if addr.TargetRef == nil || addr.TargetRef.Kind != "Pod" {
+				continue
+			}
+
+			if hostname == addr.Hostname {
+				podName := addr.TargetRef.Name
+				podNamespace := addr.TargetRef.Namespace
+				return k8sAPI.Pod().Lister().Pods(podNamespace).Get(podName)
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("no pod found in Endpoints %s/%s", svcID.Namespace, svcID.Name)
```

Comment:
perhaps include the hostname in this error message.  The Endpoints may have pods, but none may have the requested hostname.